### PR TITLE
Docstrings for helpers.py

### LINKED ISSUE #36: Docstrings for helpers.py
**Describe the bug**
The `helpers.py` file does not have clear docstrings for functions. Documentation of this would help developers have a clearer picture of the code.

**To Reproduce**
1. Check the [helpers.py](https://github.com/Praful932/Kitabe/blob/master/mainapp/helpers.py) file

**Expected behavior**
The docstrings of the functions should be documented according to`numpy style` guide.

**Screenshots**
Currently it only contains a docstring containing a short description of the function.
image

**Additional context**
For numpy style docstrings refer [this](https://numpydoc.readthedocs.io/en/latest/format.html).

**Definition of Done**
All functions are documented using numpy docstring and running this `docstr-coverage mainapp/helpers.py` command results in 100% coverage.

### PR DESCRIPTION
**Pull request by noviicee**
## Description
Documented the docstrings of functions according to numpy style guide
closes #36 

## How has this been tested?
- docstr-coverage
- flake8

## GIF Before
- NA

## GIF After
- NA

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/Praful932/Kitabe/blob/master/CONTRIBUTING.md).
- [x] My changes do not edit/add any unrequired files.
- [x] My changes are covered by tests.

### CODE REVIEW

**Review by Praful932 — CHANGES_REQUESTED** [review]
@noviicee There seems to be a lot of unecessary edits due to styling, please only edit the code wherever needed only. We need docstrings only for functions and nothing else, check [this](https://github.com/Praful932/Tf-Rec/blob/main/tfrec/utils/model_utils.py) for reference on how docstrings should be and read the issue description once.
You can revert the commit and make changes and push.

**Comment by noviicee** [comment]
> @noviicee There seems to be a lot of unecessary edits due to styling, please only edit the code wherever needed only. We need docstrings only for functions and nothing else, check [this](https://github.com/Praful932/Kitabe/pull/201) for reference on how docstrings should be and read the issue description once.
> You can revert the commit and make changes and push.

I have just edited the function docstrings and made the code according to pep8 style.
Do you want me to remove the pep8 styling?

**Comment by Praful932** [comment]
> > @noviicee There seems to be a lot of unecessary edits due to styling, please only edit the code wherever needed only. We need docstrings only for functions and nothing else, check [this](https://github.com/Praful932/Kitabe/pull/201) for reference on how docstrings should be and read the issue description once.
> > You can revert the commit and make changes and push.
> 
> I have just edited the function docstrings and made the code according to pep8 style.
> Do you want me to remove the pep8 styling?

Yes, see the example I linked above.

**Comment by noviicee** [comment]
> > > @noviicee There seems to be a lot of unecessary edits due to styling, please only edit the code wherever needed only. We need docstrings only for functions and nothing else, check [this](https://github.com/Praful932/Kitabe/pull/201) for reference on how docstrings should be and read the issue description once.
> > > You can revert the commit and make changes and push.
> > 
> > 
> > I have just edited the function docstrings and made the code according to pep8 style.
> > Do you want me to remove the pep8 styling?
> 
> Yes, see the example I linked above.

Okay I'll remove the styling..btw the link forwards to the same page we are on now 😅

**Comment by Praful932** [comment]
> > > > @noviicee There seems to be a lot of unecessary edits due to styling, please only edit the code wherever needed only. We need docstrings only for functions and nothing else, check [this](https://github.com/Praful932/Kitabe/pull/201) for reference on how docstrings should be and read the issue description once.
> > > > You can revert the commit and make changes and push.
> > > 
> > > 
> > > I have just edited the function docstrings and made the code according to pep8 style.
> > > Do you want me to remove the pep8 styling?
> > 
> > 
> > Yes, see the example I linked above.
> 
> Okay I'll remove the styling..btw the link forwards to the same page we are on now 😅

Oops sorry, just updated it

**Comment by noviicee** [comment]
> > > > > @noviicee There seems to be a lot of unecessary edits due to styling, please only edit the code wherever needed only. We need docstrings only for functions and nothing else, check [this](https://github.com/Praful932/Kitabe/pull/201) for reference on how docstrings should be and read the issue description once.
> > > > > You can revert the commit and make changes and push.
> > > > 
> > > > 
> > > > I have just edited the function docstrings and made the code according to pep8 style.
> > > > Do you want me to remove the pep8 styling?
> > > 
> > > 
> > > Yes, see the example I linked above.
> > 
> > 
> > Okay I'll remove the styling..btw the link forwards to the same page we are on now 😅
> 
> Oops sorry, just updated it

Thanks!

**Comment by noviicee** [comment]
@Praful932 some functions do not have the docstrings written already, should I leave the description empty  in those cases and only edit the parameters and returns..?

**Comment by Praful932** [comment]
> @Praful932 some functions do not have the docstrings written already, should I leave the description empty in those cases and only edit the parameters and returns..?

You have to fill in that too, since that's the purpose of this issue

**Comment by noviicee** [comment]
> > @Praful932 some functions do not have the docstrings written already, should I leave the description empty in those cases and only edit the parameters and returns..?
> 
> You have to fill in that too, since that's the purpose of this issue

Okay

**Review by Praful932 — CHANGES_REQUESTED** [review]
There are still changed lines which are visible, could you revert your branch back to master then do the changes

**Comment by maneprajakta** [comment]
@noviicee if your facing issues you can reach out to us on discord.

File: mainapp/helpers.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+"""import modules. """
```
Comment (Praful932) [inline mainapp/helpers.py:1]: Only document the functions & Classes if any :)

**Review by Praful932 — CHANGES_REQUESTED** [review]
Good going, still needs some work
checks are failing :/
you have extra spaces and empty description of parameters

File: mainapp/helpers.py (hunk 1 above)
Comment (noviicee) [reply mainapp/helpers.py:1]: It won't show 100% coverage without this line..it asks to add docstring for the module as well :/

**Comment by noviicee** [comment]
> Good going, still needs some work
> checks are failing :/
> you have extra spaces and empty description of parameters

Yup. Will remove the extra whitespaces and add description of the parameters. Though I hope there isn't a problem with a test output statement *line too long*

File: mainapp/helpers.py (hunk 1 above)
Comment (Praful932) [reply mainapp/helpers.py:1]: Oh, that's alright, ~90%+ coverage is good enough as long as methods & classes are documented

File: mainapp/helpers.py [hunk 2]
```diff
@@ -37,6 +37,22 @@
...
 def is_rating_invalid(rating):
+    """Return a boolean value.
+
+    Corresponding to whether the rating is valid.
+
+    Parameters
+    ----------
+    rating :  int
+        Rating of a book, which should be a digit <= 5.
+
```
Comment (Praful932) [inline mainapp/helpers.py:48]: prefer to leave only one line

File: mainapp/helpers.py [hunk 3]
```diff
@@ -80,25 +144,61 @@ def get_rated_bookids(user_ratings):
...
-        Returns top genre books according to a cutoff percentile to be listed in Top Books
-    '''
+    """Return top genre books according to a cutoff percentile to be listed in Top Books.
+
+    Parameters
+    ----------
+    genre : str
+        Genre of the book in string format.
+
+    percentile :    float
```
Comment (Praful932) [inline mainapp/helpers.py:192]:
```suggestion
    percentile : float
```

File: mainapp/helpers.py [hunk 4]
```diff
@@ -191,8 +345,18 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
...
+
+    Parameters
+    ----------
+    top_n : int
+         (Default value = 400)
+         Number of samples to be returned.
+
+    Returns
+    -------
+        Sample of top N books.
```
Comment (Praful932) [inline mainapp/helpers.py:358]: You can add the return type here

**Review by Praful932 — CHANGES_REQUESTED** [review]
Added some general comments, didn't add it everywhere where the issue was there to avoid repetition

File: mainapp/helpers.py [hunk 5]
```diff
@@ -37,6 +37,21 @@
 
 
 def is_rating_invalid(rating):
+    """Return a boolean value.
+
+    Corresponding to whether the rating is valid.
```
Comment (Praful932) [inline mainapp/helpers.py:42]:
```suggestion
    Checks if the rating is invalid.
```

File: mainapp/helpers.py [hunk 6]
```diff
@@ -37,6 +37,21 @@
...
+
+    Parameters
+    ----------
+    rating : int
+        Rating of a book, which should be a digit <= 5.
+
+    Returns
+    -------
+    bool
+        True if the rating is invalid, else False.
```
Comment (Praful932) [inline mainapp/helpers.py:52]:
```suggestion
        `True` if the rating is invalid, else `False`.
```

File: mainapp/helpers.py [hunk 7]
```diff
@@ -45,6 +60,21 @@ def is_rating_invalid(rating):
 
 
 def is_bookid_invalid(bookid):
+    """Return a boolean value.
+
+    Corresponding to whether the bookid is valid.
```
Comment (Praful932) [inline mainapp/helpers.py:65]:
```suggestion
    Corresponding to whether the bookid is invalid or not.
```

File: mainapp/helpers.py [hunk 8]
```diff
@@ -45,6 +60,21 @@ def is_rating_invalid(rating):
...
+
+    Parameters
+    ----------
+    bookid : int
+        book-id of the book to be checked for existence.
+
+    Returns
+    -------
+    bool
+        True if the bookid exists, else False.
```
Comment (Praful932) [inline mainapp/helpers.py:75]:
```suggestion
        `True` if the bookid exists, else `False`.
```

File: mainapp/helpers.py [hunk 9]
```diff
@@ -54,24 +84,54 @@ def is_bookid_invalid(bookid):
...
+    """Return book title given bookid.
+
+    Parameters
+    ----------
+    bookid : int
+        book-id of a book whose title needs to be determined.
+
+    Returns
+    -------
+    df : pandas.core.frame.DataFrame
```
Comment (Praful932) [inline mainapp/helpers.py:96]: The return value here is the `bookname` type str

File: mainapp/helpers.py [hunk 10]
```diff
@@ -80,25 +140,59 @@ def get_rated_bookids(user_ratings):
...
+    """Return raw_id given book_id.
+
+    Parameters
+    ----------
+    book_id : int
+        Integer to determine the raw-id of a book.
+
+    Returns
+    -------
+    df : pandas.core.frame.DataFrame
```
Comment (Praful932) [inline mainapp/helpers.py:152]: Here returned value is `raw_id` - integer

File: mainapp/helpers.py [hunk 11]
```diff
@@ -80,25 +140,59 @@ def get_rated_bookids(user_ratings):
...
+    """
     bookid_list = list(df_book[df_book.r_index.isin(raw_id_list)]['book_id'].values)
     return bookid_list
 
 
 def genre_wise(genre, percentile=0.85):
-    '''
-        Returns top genre books according to a cutoff percentile to be listed in Top Books
-    '''
+    """Return top genre books according to a cutoff percentile to be listed in Top Books.
```
Comment (Praful932) [inline mainapp/helpers.py:179]:
```suggestion
    """Return top genre books according to a cutoff percentile.
```

File: mainapp/helpers.py [hunk 12]
```diff
@@ -80,25 +140,59 @@ def get_rated_bookids(user_ratings):
...
+        Genre of the book in string format.
+
+    percentile : float
+         (Default value = 0.85)
+         Float determinig the cutoff percentile.
+
+    Returns
+    -------
+    df : pandas.core.frame.DataFrame
+        Top genre books to be listed in Top Books according to a cutoff percentile.
```
Comment (Praful932) [inline mainapp/helpers.py:193]:
```suggestion
        Top genre books according to a cutoff percentile.
```

File: mainapp/helpers.py [hunk 13]
```diff
@@ -80,25 +140,59 @@ def get_rated_bookids(user_ratings):
...
+    """Return top genre books according to a cutoff percentile to be listed in Top Books.
+
+    Parameters
+    ----------
+    genre : str
+        Genre of the book in string format.
+
+    percentile : float
+         (Default value = 0.85)
+         Float determinig the cutoff percentile.
```
Comment (Praful932) [inline mainapp/helpers.py:188]:
```suggestion
         Float determinig the cutoff percentile (Default value = `0.85`).
```

File: mainapp/helpers.py [hunk 14]
```diff
@@ -116,9 +210,19 @@ def genre_wise(genre, percentile=0.85):
 
 
 def count_vectorizer_recommendations(bookid):
-    '''
-        Returns recommened book ids based on book details
-    '''
+    """Return recommened book ids based on book details.
```
Comment (Praful932) [inline mainapp/helpers.py:213]:
```suggestion
    """Return recommened based on count vectorizer.
```

File: mainapp/helpers.py [hunk 15]
```diff
@@ -191,8 +336,19 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
...
 def get_top_n(top_n=400):
-    """
-        Returns a sample of top N books
+    """Return a sample of top N books.
+
+    Parameters
+    ----------
+    top_n : int
+         (Default value = 400)
+         Number of samples to be returned.
```
Comment (Praful932) [inline mainapp/helpers.py:345]:
```suggestion
         Number of samples to be returned (Default value = 400).
```

File: mainapp/helpers.py [hunk 16]
```diff
@@ -207,12 +363,24 @@ def get_top_n(top_n=400):
...
+
         If enough books are not available, top books are
         sampled randomly.
-    '''
+
+    Parameters
+    ----------
+    N : int
+         (Default value = 15)
+         Number of samples to be returned.
```
Comment (Praful932) [inline mainapp/helpers.py:375]: Same issue here

**Review by Praful932 — CHANGES_REQUESTED** [review]
Added some comments, a general tip, if the variable name is not mentioned for the return type you can add a variable name by looking at the context of the function for instance
```
dict
        Dictionary of book details.
```
```
book_details : dict
        Dictionary of book details.
```

**Comment by Praful932** [comment]
When you have resolved the comments, you can notify us by clicking on `request review`
Thanks

File: mainapp/helpers.py [hunk 17]
```diff
@@ -80,25 +140,58 @@ def get_rated_bookids(user_ratings):
...
+
+    Parameters
+    ----------
+    book_id : int
+        Integer to determine the raw-id of a book.
+
+    Returns
+    -------
+    raw_id : int
+        Dataframe of shape `(1,1)` having the raw-id.
```
Comment (Praful932) [inline mainapp/helpers.py:153]:
```suggestion
        corresponding raw_id of the book_id.
```

File: mainapp/helpers.py [hunk 18]
```diff
@@ -116,9 +209,19 @@ def genre_wise(genre, percentile=0.85):
 
 
 def count_vectorizer_recommendations(bookid):
-    '''
-        Returns recommened book ids based on book details
-    '''
+    """Return recommened based on count vectorizer.
```
Comment (Praful932) [inline mainapp/helpers.py:212]:
```suggestion
    """Return recommenedations based on count vectorizer.
```

File: mainapp/helpers.py [hunk 19]
```diff
@@ -163,19 +276,49 @@ def embedding_recommendations(sorted_user_ratings):
...
+    already_rated : list
+        List containing book-ids of already rated books.
+
+    recommendations : int
+         Integer denoting the number of recommendations (Default value = 9).
+
+    Returns
+    -------
+    best_bookids : list
+        List containing bookids of top books based on embeddings, ratings, and book details.
```
Comment (Praful932) [inline mainapp/helpers.py:319]:
```suggestion
        List containing bookids of top books based on embeddings and countvectorizer.
```

File: mainapp/helpers.py [hunk 20]
```diff
@@ -191,8 +334,18 @@ def combine_ids(cv_bookids, embedding_bookids, already_rated, recommendations=9)
 
 
 def get_top_n(top_n=400):
-    """
-        Returns a sample of top N books
+    """Return a sample of top N books.
```
Comment (Praful932) [inline mainapp/helpers.py:337]:
```suggestion
    """Return a sample of top N books based on weighted average ratings.
```

File: mainapp/helpers.py [hunk 21]
```diff
@@ -207,12 +360,23 @@ def get_top_n(top_n=400):
...
+    ----------
+    N : int
+         Number of samples to be returned (Default value = 15).
+
+    Returns
+    -------
+    book_details : dict
+        Dictionary of book details.
+
+        Value based on the value returned from the called function.
```
Comment (Praful932) [inline mainapp/helpers.py:378]:
```suggestion
```

**Review by Praful932 — CHANGES_REQUESTED** [review]
Added some comments, could you add a ss of `doc coverage` of this file, then it would be good to go

**Review by Praful932 — APPROVED** [review]
LGTM 🚀

**Review by maneprajakta — APPROVED** [review]
Great Work!

**Comment by noviicee** [comment]
> Great Work!

Thanks a lot!

**Pull request merged**